## Lab 4 - Practical Lab 4 - Multivariate Linear and Polynomial Regression, and Evaluation using R-Squared, MAPE and MAE
## Name: Nihal Patel
## Student ID: 8945100

In this assignment, we will run multivariate linear model and two polynomial regression models, and evaluate them using R-squared, Mean Absolute Percentage Error (MAPE) and Mean Absolute Error (MAE) metrices. The dataset will be Scikit-Learn's Diabetes dataset.

In [27]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score, mean_absolute_error, mean_absolute_percentage_error

# 1. Get the data, and run a train-validation-test split. Description of each column can be found in sklearn documentation. Look at the documentation for the load_diabetes method to know what are as_frame and scaled arguments are for.

In [28]:
# We will load the diabetes data set
diabetes_X,diabetes_y = load_diabetes(return_X_y=True, as_frame=True, scaled=True)

In [29]:
# We will split the data into training, testing and validation set
diabetes_X_train, diabetes_X_test, diabetes_y_train, diabetes_y_test = train_test_split(diabetes_X,diabetes_y,test_size=0.2, random_state=40)
diabetes_X_train, diabetes_X_val, diabetes_y_train, diabetes_y_val = train_test_split(diabetes_X,diabetes_y,test_size=0.25, random_state=40)

Here, we are splitting the data into 80% train and 20% test and then we split the train set and validation into 75% and 25%.

# 2. Run a multivariate linear regression on all variables (1 point)

In [30]:
#Now, we will run a multivariate linear regression on all the variables which are there in the dataset using LinearRegression class from sklearn.
#Here, we will fit the model on the train set and evaluation will be done on the validation set using R-sqaured, MAE and MAPE.

# We will run a multivariate linear regression on all variables
lin_reg = LinearRegression()
lin_reg.fit(diabetes_X_train, diabetes_y_train)

LinearRegression()

In [31]:
# We will do prediction on the validation set
y_pred_Linear = lin_reg.predict(diabetes_X_val)

In [32]:
# We will evaluate the model using R-squared, MAPE and MAE
r2linear = r2_score(diabetes_y_val, y_pred_Linear)
MAPE_linear = mean_absolute_percentage_error(diabetes_y_val, y_pred_Linear)
MAE_linear = mean_absolute_error(diabetes_y_val, y_pred_Linear)

# Printing the results
print(f"Multivariate linear regression results are as below:\n")
print(f"R-squared: {r2linear:.2f}\n")
print(f"MAPE: {MAPE_linear:.2f}\n")
print(f"MAE: {MAE_linear:.2f}\n")

Multivariate linear regression results are as below:

R-squared: 0.39

MAPE: 0.42

MAE: 50.93



# 3. Run a polynomial regression of the 2nd degree on the BMI feature alone

In [33]:
#Further, we will perform polynomial regression of 2nd degree on the BMI feature using the PolynomialFeatures class from sklearn.
#We will be transforming the BMI feature into a quadratic polynomial and will fit a linear regression on it.
#Running a polynomial regression of the 2nd degree on the BMI feature alone
poly_bmi = PolynomialFeatures(degree=2)
diabetes_X_train_bmi = poly_bmi.fit_transform(diabetes_X_train[["bmi"]])
diabetes_X_val_bmi = poly_bmi.transform(diabetes_X_val[["bmi"]])

poly_reg_bmi = LinearRegression()
poly_reg_bmi.fit(diabetes_X_train_bmi, diabetes_y_train)

LinearRegression()

In [34]:
# We will perform Prediction on the validation set
y_pred_poly_bmi = poly_reg_bmi.predict(diabetes_X_val_bmi)

In [35]:
# Now we will evaluate the model using R-squared, MAPE and MAE
r2_poly_bmi = r2_score(diabetes_y_val, y_pred_poly_bmi)
MAPE_poly_bmi = mean_absolute_percentage_error(diabetes_y_val, y_pred_poly_bmi)
MAE_poly_bmi = mean_absolute_error(diabetes_y_val, y_pred_poly_bmi)

# Printing the results
print(f"Polynomial regression on BMI results:\n")
print(f"R-squared: {r2_poly_bmi:.2f}\n")
print(f"MAPE: {MAPE_poly_bmi:.2f}\n")
print(f"MAE: {MAE_poly_bmi:.2f}\n")

Polynomial regression on BMI results:

R-squared: 0.25

MAPE: 0.47

MAE: 58.72



# 4. Run a multivariate polynomial regression of the 2nd degree on all variables (Hint: set include_bias=False in PolynomialFeatures)

In [36]:
#Now we will run a multivariate polynomial regression of the 2nd degree on all variables using PolynomialFeatures class from sklearn.
#All features will be transformed into quadratic polynomials and fit a linear regression on them.

# We will now run a multivariate polynomial regression of the 2nd degree on all variables
poly_reg_all = PolynomialFeatures(degree=2, include_bias=False)
diabetes_X_train_all = poly_reg_all.fit_transform(diabetes_X_train)
diabetes_X_val_all = poly_reg_all.transform(diabetes_X_val)

poly_reg_all = LinearRegression()
poly_reg_all.fit(diabetes_X_train_all, diabetes_y_train)

LinearRegression()

In [37]:
# We will perform Prediction on the validation set
y_pred_poly_all = poly_reg_all.predict(diabetes_X_val_all)

In [38]:
# We will evaluate the model using R-squared, MAPE and MAE
r2_poly_all = r2_score(diabetes_y_val, y_pred_poly_all)
MAPE_poly_all = mean_absolute_percentage_error(diabetes_y_val, y_pred_poly_all)
MAE_poly_all = mean_absolute_error(diabetes_y_val, y_pred_poly_all)

# Printing the results
print(f"Multivariate polynomial regression results:\n")
print(f"R-squared: {r2_poly_all:.2f}\n")
print(f"MAPE: {MAPE_poly_all:.2f}\n")
print(f"MAE: {MAE_poly_all:.2f}\n")

Multivariate polynomial regression results:

R-squared: 0.22

MAPE: 0.45

MAE: 56.50



# 5. Compare the three models by looking at R-squared, MAPE and MAE. Explain what the values mean for a non-expert and add your insight about the values of each model. Note: You can add any further comparisons and code (this is not necessary for a perfect score, but will be reviewed and evaluated)

We use R-square method to check how well the model fits the data where it ranges from 0 to 1, where the higher values indicate us about a better fit.
MAPE and MAE are measures which are used to show how much the model deviates from the true values. They both are in the same units as the target variable where lower values indicate better accuracy.

A non-expert can be explained about the values in the below given way:
The multivariate linear regression model explains about 39% of the variation in the target variable, and has an average error of about 51 units. The polynomial regression on BMI model explains about 25% of the variation in the target variable, and has an average error of about 59 units. Whereas, the multivariate polynomial regression model explains about 22% of the variation in the target variable, and has an average error of about 56 units.

I have some insights for the values of each model mentioned as below:
1. Multivariate Linear Regression model performs much better than the Polynomial and Multivariate Polynomial Regression models.
2. It performs well because of the linear relationship of the target variables and its characteristics which differ from other two models.
3. Polynomial Regression is seen performing worst on BMI Model which can be considered as a bad target variable predictor. 
4. The multivariate polynomial regression model performs better than the polynomial regression on BMI model but not good than the multivariate linear regression model. It will indicate that the nonlinear components do not improvise the model's fit or accuracy.

# 6. Please answer the following questions:

a. How many parameters are we fitting for each of the three models? Explain these values. Hint: for explaining the parameters of the polynomial regression, you can use poly.get_feature_names_out()

For every regression model, we are fitting parameters as below:
1. Multivariate Linear Regression Model : Total of 11 parameters which contain an intercept and a coefficient for each 10 features.
2. Polynomial Regression on BMI Model: Total of 3 parameters which contain an intercept and two coefficients for the quadratic polynomial of BMI.
3. Multivariate Polynomial Regression Model: Total of 66 parameters which contain an intercept and 65 coefficients for the quadratic polynomials of all other features.

We can get the above given number of coefficients by using poly_all.get_feature_names_out(). It will display that there are 10 linear terms, 45 interaction terms, and 10 quadratic terms.

b. Which model would you choose for deployment, and why?

I would give preference to the Multivariate Linear Regression model for deployment considering its highest R-squared value and the lowest MAPE and MAE values amongst the three models. It also means that it has the best fit and accuracy for predicting the target variable. But before concluding, we can also consider the other factors like its complexity, interpretability and generalization.